core/cap-0072.md

```diff
@@ -0,0 +1,341 @@
++    ACCOUNT_MERGE_IS_SPONSOR = -7,      // can't merge account that is a sponsor
++    // There is an additional signers entry for this account that has been 
++    // archived. It has to be restored first before account can be merged.
++    ACCOUNT_MERGE_ADDITIONAL_SIGNERS_ARCHIVED = -8 
+ };
+ 
+ union AccountMergeResult switch (AccountMergeResultCode code)
+```
+
+### Stellar(G-) account contract interface
+
+This CAP introduces a callable smart contract interface for the Stellar accounts. The following Rust trait specifies the interface as Soroban Rust SDK contract.
+
+```rust
+pub trait StellarAccountInterface {
+    /// Adds a new ed25519 signer to the account with the given weight, or
+    /// updates the weight if the signer already exists.
+    /// 
+    /// Weight must be in range (0, 255].
+    fn update_ed25519_signer(env: Env, key: BytesN<32>, weight: u32);
+
+    /// Removes an ed25519 signer from the account.    
+    fn remove_ed25519_signer(env: Env, key: BytesN<32>);
+
+    /// Adds a new delegated signer to the account with the given weight, or
+    /// updates the weight if the signer already exists.
+    /// 
+    /// Weight must be in range (0, 255].
+    ///
+    /// Instead of performing signature verification for the delegated signers,
+    /// their authentication logic (`__check_auth`) will be called instead via
+    /// `delegate_account_auth` host function.
+    fn update_delegated_signer(env: Env, signer: Address, weight: u32);
+
+    /// Removes a delegated signer from the account.
+    fn remove_delegated_signer(env: Env, signer: Address);
+
+    /// Sets the weight for the 'master' key of the account.
+    /// 
+    /// The 'master' key is the public ed25519 key that identifies the account
+    /// itself. Setting the weight to 0 effectively removes the master key
+    /// from the account.
+    fn set_master_weight(env: Env, weight: u32);
+
+    /// Updates the signature thresholds for the account.
+    /// 
+    /// `None` values leave the corresponding threshold unchanged.
+    /// 
+    /// `low` threshold is only used for authorizing a few non-sensitive Stellar
+    //  operations, such as bumping the account's sequence number.
+    /// 
+    /// `med` threshold is used for authorizing most of the operations, 
+    /// including the Smart Contract interactions authorized via `require_auth`.
+    /// 
+    /// `high` threshold is used for managing the account itself. Note, that
+    /// all the functions in this interface require high threshold, while still
+    /// being authorized via `require_auth`.
+    fn update_thresholds(env: Env, low: Option<u32>, med: Option<u32>, high: Option<u32>);
+}
+```
+
+### 'Classic' transaction semantics
+
+#### Ledger entry for additional account signers
+
+Every new or existing G-account on-chain may have up to 1 `ContractDataEntry` that contains additional signers (further referred to as 'additional signers entry'). The specification for the entry is as follows:
```

Comment:
Some clarifying questions - 
1. Can classic operations use these signers?
2. Does this double the number of total signers allowed for a G account from 20 to 40?
3. If `SetOptionsOp` tries to add a signer that exists in the additional signers entry, it should update the weight of the additional signers right? What if the additional signers entry isn't live? If you added that signer and then restored, they would exist in both places.

Reply:
After updating this I feel like there might be too many exceptions for this option to be better than adding a new signer type to the account  and dealing with base reserves and sponsorships on the soroban side. I'd still merge the CAP as is, and think about the alternative for a bit.

Reply:
> No, because C-signers are only allowed in smart contract environment as per CAP (though if soroban environment is ever supported for the classic ops, then the additional C-signers for G-accounts would automatically work as well).

Sorry my question wasn't clear. I'm asking if the new ed25519 signers are used in classic authorization.


Also, we should merge this CAP as is. I just thought it was easier to ask questions in the PR, but that shouldn't block the merge.

Reply:
> Sorry my question wasn't clear. I'm asking if the new ed25519 signers are used in classic authorization.

They are, that's documented in the classic semantics section. E.g. `Transaction signature verification rules update` describes the signature validation flow changes.